fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/InferPredicates.java

```diff
@@ -0,0 +1,137 @@
+            root.getConjuncts().forEach(filterPredicates::remove);
+            if (!filterPredicates.isEmpty()) {
+                filterPredicates.addAll(root.getConjuncts());
+                return new LogicalFilter<>(ExpressionUtils.and(Lists.newArrayList(filterPredicates)), root.child());
+            }
+            return root;
+        }).toRule(RuleType.INFER_PREDICATES_FOR_WHERE);
+    }
+
+    /**
+     * The logic is as follows:
+     * 1. poll up bottom predicate then infer additional predicates
+     *   for example:
+     *   select * from (select * from t1 where t1.id = 1) t join t2 on t.id = t2.id
+     *   1. poll up bottom predicate
+     *      select * from (select * from t1 where t1.id = 1) t join t2 on t.id = t2.id and t.id = 1
+     *   2. infer
+     *      select * from (select * from t1 where t1.id = 1) t join t2 on t.id = t2.id and t.id = 1 and t2.id = 1
+     *   finally transformed sql:
+     *      select * from (select * from t1 where t1.id = 1) t join t2 on t.id = t2.id and t2.id = 1
+     * 2. put these predicates into `otherJoinConjuncts` , these predicates are processed in the next
+     *   round of predicate push-down
+     */
+    private Rule inferOn() {
+        return logicalJoin(any(), any()).thenApply(ctx -> {
+            LogicalJoin<Plan, Plan> root = ctx.root;
+            Plan left = getOriginalPlan(ctx, root.left());
+            Plan right = getOriginalPlan(ctx, root.right());
+            Set<Expression> expressions = getAllExpressions(left, right, root.getOnClauseCondition());
+            List<Expression> otherJoinConjuncts = Lists.newArrayList(root.getOtherJoinConjuncts());
+            switch (root.getJoinType()) {
+                case INNER_JOIN:
+                case CROSS_JOIN:
+                case LEFT_SEMI_JOIN:
+                case RIGHT_SEMI_JOIN:
+                    otherJoinConjuncts.addAll(inferNewPredicate(left, expressions));
+                    otherJoinConjuncts.addAll(inferNewPredicate(right, expressions));
+                    break;
+                case LEFT_OUTER_JOIN:
+                case LEFT_ANTI_JOIN:
+                    otherJoinConjuncts.addAll(inferNewPredicate(right, expressions));
+                    break;
+                case RIGHT_OUTER_JOIN:
+                case RIGHT_ANTI_JOIN:
+                    otherJoinConjuncts.addAll(inferNewPredicate(left, expressions));
+                    break;
+                default:
+                    return root;
+            }
+            return root.withOtherJoinConjuncts(otherJoinConjuncts);
+        }).toRule(RuleType.INFER_PREDICATES_FOR_ON);
+    }
+
+    private Plan getOriginalPlan(MatchingContext context, Plan patternPlan) {
+        patternPlan.getGroupExpression().orElseThrow(() -> new IllegalArgumentException("GroupExpression not exists."));
+        return context.cascadesContext.getMemo().copyOut(patternPlan.getGroupExpression().get(), false);
+    }
```

Comment:
This copy out inefficiency, every filter/join will copy out every child plan tree, there are a lot of duplicate copies.
How about copy out the root group once, and do bottom up rewrite, then copy into the memo